components/renku_data_services/utils/core.py

```diff
@@ -69,3 +71,30 @@ async def transaction_wrapper(self: WithSessionMaker, *args, **kwargs):
             return await f(self, session, *args, **kwargs)
 
     return transaction_wrapper
+
+
+P = ParamSpec("P")
+T = TypeVar("T")
+_WithSessionMaker = TypeVar("_WithSessionMaker", bound=WithSessionMaker)
+_WithTransactionFunc: TypeAlias = Callable[Concatenate[_WithSessionMaker, P], Awaitable[T]]
+
+
+def with_db_transaction(f: _WithTransactionFunc) -> _WithTransactionFunc:
+    """Initializes a transaction and commits it on successful exit of the wrapped function."""
+
+    @functools.wraps(f)
+    async def transaction_wrapper(self: _WithSessionMaker, *args: P.args, **kwargs: P.kwargs):
+        session_kwarg = kwargs.get("session")
+        if "session" in kwargs and session_kwarg is not None and not isinstance(session_kwarg, AsyncSession):
+            raise errors.ProgrammingError(
+                message="The decorator that starts a DB transaction expects the session keyword to be "
+                "present in the function arguments"
```

Comment:
I think this message is wrong, it expects it to not be present or to be a valid session?